kernel: various kernel module build fixes for the upcoming 6.12 kernel

DragonBluep opened the pull request:
When I tried to build the 6.12 kernel pull request, I found that there were still many compilation errors. This PR fixes the kernel build errors that I can catch.

namiltd commented:
Check out https://github.com/namiltd/openwrt/pull/65/commits Some of the fixes there overlap with yours.

DragonBluep committed:
generic: fix disable common USB quirks patch

Remove modifications for codes guarded by CONFIG_USB_PCI_AMD.
This kernel symbol is only visible on desktop level platform,
e.g. x86. They are not sensitive to firmware image size. This
patch also fixes function redefinition build errors:

drivers/usb/host/pci-quirks.c:621:6: error: redefinition of 'usb_asmedia_modifyflowcontrol'
  621 | void usb_asmedia_modifyflowcontrol(struct pci_dev *pdev)
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from drivers/usb/host/pci-quirks.c:22:
drivers/usb/host/pci-quirks.h:49:20: note: previous definition of 'usb_asmedia_modifyflowcontrol' with type 'void(struct pci_dev *)'
   49 | static inline void usb_asmedia_modifyflowcontrol(struct pci_dev *pdev) {}
      |                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Fixes: be6753dda00c ("generic: 6.12: manually rebuild pending patches")
Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18637
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

KanjiMonster commented on package/kernel/mac80211/patches/ath9k/551-ath9k_ubnt_uap_plus_hsr.patch:
```diff
@@ -308,6 +309,8 @@
 +#ifndef HSR_H
 +#define HSR_H
 +
++#include "hw.h"
```
another option is to just define the struct, like .e.g `mac.h` does to avoid requiring an include:

```suggestion
+struct ath_hw;
```

The header doesn't care what ath_hw looks like, just that it is something that can be pointed at.

DragonBluep replied:
I reorganized the header files in `hsr.c` so that we can avoid including `hw.h` here. diff: https://github.com/openwrt/openwrt/compare/f434a28102960c43dc6225658007ba84e67b3792..1996a64c6e18bda6e1e3deffa1f742122eef8a16

KanjiMonster replied:
hsr.h uses `struct ath_hw`, so it should either include `hw.h` or define it itself, it shouldn't depend on users having included `hw.h` before.

This will then e.g. break if someone comes along and reorders the includes alphabetically (which isn't too uncommon).

DragonBluep replied:
Thank you for your explanation. I added `struct ath_hw;` to the header file.

1715173329 commented on package/kernel/rtl8812au-ct/Makefile:
```diff
@@ -41,6 +41,7 @@ NOSTDINC_FLAGS := \
 	-I$(STAGING_DIR)/usr/include/mac80211/uapi \
 	-include backport/backport.h \
 	-Wno-error=address \
+	-Wno-error=missing-prototypes \
```
maybe this driver can be removed, mac80211 have native support for it now (kmod-rtw88-8812au).
Cc: @hauke

hauke replied:
Maybe for some use cases this older driver works better, but I would not invest too much time into it. We should prefer the driver form the mac80211 backports.

hauke commented on target/linux/generic/pending-6.12/811-pci_disable_usb_common_quirks.patch:
```diff
@@ -20,16 +20,25 @@ Signed-off-by: Felix Fietkau <nbd@nbd.name>
+@@ -724,6 +730,10 @@ reset_needed:
  	return 1;
  }
  EXPORT_SYMBOL_GPL(uhci_check_and_reset_hc);
```
The original patch exports an empty version of `uhci_check_and_reset_hc` when `#if IS_ENABLED(CONFIG_USB_UHCI_HCD)` is false.

DragonBluep replied:
In 6.12 kernel upstream commit https://github.com/torvalds/linux/commit/358ad297e379ff548247e3e24c6619559942bfdd, `uhci_check_and_reset_hc` and `quirk_usb_handoff_uhci` are guarded by `#if IS_ENABLED(CONFIG_USB_UHCI_HCD)`. Hence `uhci_check_and_reset_hc()` is **unused** if not define `CONFIG_USB_UHCI_HCD`. We don't need to hack it again.

DragonBluep replied:
ping @Ansuel

hauke commented:
I took all patches except the last one, I think there are more problems in that USB quirk patch.

hauke commented on target/linux/generic/pending-6.12/811-pci_disable_usb_common_quirks.patch:
```diff
@@ -20,16 +20,25 @@ Signed-off-by: Felix Fietkau <nbd@nbd.name>
+ EXPORT_SYMBOL_GPL(usb_asmedia_modifyflowcontrol);
++#endif /* CONFIG_USB_PCI */
++
++#endif /* !CONFIG_PCI_DISABLE_COMMON_QUIRKS */
```
This is still overlapping with the `#ifdef CONFIG_USB_PCI_AMD`  block. The `#ifdef CONFIG_USB_PCI_AMD` is only activated on x86 and maybe longsoon, they are not space constrained we do not need extra handling in our patch for them. 

I think we only need the last block of the patch for this file, the rest is already taken care of upstream.

namiltd replied:
Take a look at my solution: https://github.com/openwrt/openwrt/pull/18720

DragonBluep replied:
> This is still overlapping with the `#ifdef CONFIG_USB_PCI_AMD` block. The `#ifdef CONFIG_USB_PCI_AMD` is only activated on x86 and maybe longsoon, they are not space constrained we do not need extra handling in our patch for them.

This makes sense. I have removed the modifications about `CONFIG_USB_PCI_AMD`.

> 
> I think we only need the last block of the patch for this file, the rest is already taken care of upstream.

We still need to exclude `usb_asmedia_modifyflowcontrol()` when `CONFIG_PCI_DISABLE_COMMON_QUIRKS` is enabled. It fixes the function redefinition error. On 6.6 kernel, `usb_asmedia_modifyflowcontrol()` is an empty function, too.
